File: apps/mobile-ios/Packages/TatameCore/Sources/TatameCore/AttendanceRepository.swift
// Repository seam for the attendance slice (spec 004, ATT.22-24). Same
// convention as EnrollmentRepository: protocol in TatameCore, implementation
// in TatameAPI, features depend only on this protocol and throw `ApiError`.
//
// `liveStream` is the one non-REST member: the SSE endpoint is a documented
// exception to the OpenAPI contract (issue 09) — the Live implementation
// hand-rolls URLSession.bytes + an SSE parser; fakes hand the model a
// controllable stream.

import Foundation

public protocol AttendanceRepository: Sendable {
    // MARK: Aluno

    /// GET /aluno/home — hero context + stat tiles + graduation numerator.
    func alunoHome() async throws -> AlunoHome

    /// POST /aluno/checkins — one write path, three methods. Exactly one of
    /// `qrToken` (qr) / `code` (code) / `classId` (manual) per the method.
    /// Duplicates come back as `.alreadyCheckedIn`, never an error.
    func checkIn(method: CheckinMethod, qrToken: String?, code: String?, classId: UUID?) async throws -> CheckinResult

    // MARK: Professor — chamada ao vivo (own classes only; foreign ids are 404s)

    /// POST /professor/classes/:id/live-codes — open chamada (idempotent:
    /// reopening while a code is active returns the active code).
    func openLiveCode(classId: UUID) async throws -> LiveCode

    /// POST /professor/live-codes/:id/close — encerrar chamada.
    func closeLiveCode(id: UUID) async throws -> LiveCode

    /// GET /professor/live-codes/:id/attendances — snapshot + polling target.
    func liveSnapshot(liveCodeId: UUID) async throws -> LiveSnapshot

    /// POST /professor/live-codes/:id/stream-ticket — ~60 s signed ticket.
    func mintStreamTicket(liveCodeId: UUID) async throws -> StreamTicket

    /// GET /professor/live-codes/:id/stream?ticket= — named `checkin` /
    /// `revoke` events; heartbeats are transport noise the stream never
    /// surfaces. Finishes (throwing) when the connection drops.
    func liveStream(liveCodeId: UUID, ticket: String) -> AsyncThrowingStream<LiveStreamEvent, any Error>

    // MARK: Professor — chamada manual

    /// POST /professor/classes/:id/roll-call — open manual chamada; self
    /// check-ins arrive pre-toggled.
    func openRollCall(classId: UUID) async throws -> RollCall

    /// POST /professor/sessions/:id/attendances — toggle on (method manual,
    /// recorded by the professor).
    func markAttendance(sessionId: UUID, studentId: UUID) async throws -> MarkAttendanceResult

    /// POST /professor/attendances/:id/revoke — same-day toggle-off.
    /// Throws `.forbidden(attendance.revoke_window_closed)` after day close.
    func revokeAttendance(id: UUID) async throws -> RevokeAttendanceResult

    // MARK: Professor — dashboard & students

    /// GET /professor/dashboard — alunos hoje, presença média, next-class hero.
    func dashboard() async throws -> ProfessorDashboard

    /// GET /professor/students?notEnrolledInClassId= — the "Adicionar aluno"
    /// picker source (closes the spec-003 API-gap workaround).
    func students(notEnrolledInClassId: UUID?) async throws -> [RosterStudent]
}

/// Default used by environment entries / previews; every call traps.
public struct UnimplementedAttendanceRepository: AttendanceRepository {
    public init() {}

    public func alunoHome() async throws -> AlunoHome {
        fatalError("AttendanceRepository not injected")
    }

    public func checkIn(
        method _: CheckinMethod,
        qrToken _: String?,
        code _: String?,
        classId _: UUID?
    ) async throws -> CheckinResult {
        fatalError("AttendanceRepository not injected")
    }

    public func openLiveCode(classId _: UUID) async throws -> LiveCode {
        fatalError("AttendanceRepository not injected")
    }

    public func closeLiveCode(id _: UUID) async throws -> LiveCode {
        fatalError("AttendanceRepository not injected")
    }

    public func liveSnapshot(liveCodeId _: UUID) async throws -> LiveSnapshot {
        fatalError("AttendanceRepository not injected")
    }

    public func mintStreamTicket(liveCodeId _: UUID) async throws -> StreamTicket {
        fatalError("AttendanceRepository not injected")
    }

    public func liveStream(liveCodeId _: UUID, ticket _: String) -> AsyncThrowingStream<LiveStreamEvent, any Error> {
        fatalError("AttendanceRepository not injected")
    }

    public func openRollCall(classId _: UUID) async throws -> RollCall {
        fatalError("AttendanceRepository not injected")
    }

    public func markAttendance(sessionId _: UUID, studentId _: UUID) async throws -> MarkAttendanceResult {
        fatalError("AttendanceRepository not injected")
    }

    public func revokeAttendance(id _: UUID) async throws -> RevokeAttendanceResult {
        fatalError("AttendanceRepository not injected")
    }

    public func dashboard() async throws -> ProfessorDashboard {
        fatalError("AttendanceRepository not injected")
    }

    public func students(notEnrolledInClassId _: UUID?) async throws -> [RosterStudent] {
        fatalError("AttendanceRepository not injected")
    }
}

File: apps/mobile-android/app/src/main/kotlin/br/com/tatame/core/attendance/AttendanceRepository.kt
package br.com.tatame.core.attendance

import br.com.tatame.core.network.ApiResult
import br.com.tatame.core.network.AttendanceApi
import br.com.tatame.core.network.ProblemJson
import br.com.tatame.core.network.apiCall
import br.com.tatame.core.network.dto.AlunoHomeResponse
import br.com.tatame.core.network.dto.CheckinMethods
import br.com.tatame.core.network.dto.CheckinRequest
import br.com.tatame.core.network.dto.CheckinResponse
import br.com.tatame.core.network.dto.LiveCodeResponse
import br.com.tatame.core.network.dto.LiveSnapshotResponse
import br.com.tatame.core.network.dto.MarkAttendanceRequest
import br.com.tatame.core.network.dto.MarkAttendanceResponse
import br.com.tatame.core.network.dto.ProfessorDashboardResponse
import br.com.tatame.core.network.dto.ProfessorStudent
import br.com.tatame.core.network.dto.RevokeAttendanceRequest
import br.com.tatame.core.network.dto.RevokeAttendanceResponse
import br.com.tatame.core.network.dto.RollCallResponse
import br.com.tatame.core.network.dto.StreamTicketResponse
import br.com.tatame.core.network.map
import kotlinx.serialization.json.Json

/**
 * Seam the attendance feature ViewModels talk through (fakeable in JVM
 * tests) — same convention as [br.com.tatame.core.enrollment.EnrollmentRepository].
 * One repository method per check-in method keeps the request-shape rule
 * (`method` + its single payload field) in one place.
 */
interface AttendanceRepository {
    // aluno (ATT.19)
    suspend fun checkInQr(qrToken: String): ApiResult<CheckinResponse>
    suspend fun checkInCode(code: String): ApiResult<CheckinResponse>
    suspend fun checkInManual(classId: String): ApiResult<CheckinResponse>
    suspend fun alunoHome(): ApiResult<AlunoHomeResponse>

    // professor chamada ao vivo (ATT.20)
    suspend fun openLiveCode(classId: String): ApiResult<LiveCodeResponse>
    suspend fun closeLiveCode(liveCodeId: String): ApiResult<LiveCodeResponse>
    suspend fun liveSnapshot(liveCodeId: String): ApiResult<LiveSnapshotResponse>
    suspend fun mintStreamTicket(liveCodeId: String): ApiResult<StreamTicketResponse>

    // professor chamada manual (ATT.21)
    suspend fun openRollCall(classId: String): ApiResult<RollCallResponse>
    suspend fun markAttendance(sessionId: String, studentId: String): ApiResult<MarkAttendanceResponse>
    suspend fun revokeAttendance(attendanceId: String): ApiResult<RevokeAttendanceResponse>

    // professor dashboard & students (ATT.21)
    suspend fun dashboard(): ApiResult<ProfessorDashboardResponse>
    suspend fun students(notEnrolledInClassId: String? = null): ApiResult<List<ProfessorStudent>>
}

class AttendanceRepositoryImpl(
    private val api: AttendanceApi,
    private val json: Json = ProblemJson,
) : AttendanceRepository {

    override suspend fun checkInQr(qrToken: String): ApiResult<CheckinResponse> =
        apiCall(json) { api.checkIn(CheckinRequest(method = CheckinMethods.QR, qrToken = qrToken)) }

    override suspend fun checkInCode(code: String): ApiResult<CheckinResponse> =
        apiCall(json) { api.checkIn(CheckinRequest(method = CheckinMethods.CODE, code = code)) }

    override suspend fun checkInManual(classId: String): ApiResult<CheckinResponse> =
        apiCall(json) { api.checkIn(CheckinRequest(method = CheckinMethods.MANUAL, classId = classId)) }

    override suspend fun alunoHome(): ApiResult<AlunoHomeResponse> =
        apiCall(json) { api.alunoHome() }

    override suspend fun openLiveCode(classId: String): ApiResult<LiveCodeResponse> =
        apiCall(json) { api.openLiveCode(classId) }

    override suspend fun closeLiveCode(liveCodeId: String): ApiResult<LiveCodeResponse> =
        apiCall(json) { api.closeLiveCode(liveCodeId) }

    override suspend fun liveSnapshot(liveCodeId: String): ApiResult<LiveSnapshotResponse> =
        apiCall(json) { api.liveSnapshot(liveCodeId) }

    override suspend fun mintStreamTicket(liveCodeId: String): ApiResult<StreamTicketResponse> =
        apiCall(json) { api.mintStreamTicket(liveCodeId) }

    override suspend fun openRollCall(classId: String): ApiResult<RollCallResponse> =
        apiCall(json) { api.openRollCall(classId) }

    override suspend fun markAttendance(
        sessionId: String,
        studentId: String,
    ): ApiResult<MarkAttendanceResponse> =
        apiCall(json) { api.markAttendance(sessionId, MarkAttendanceRequest(studentId)) }

    override suspend fun revokeAttendance(attendanceId: String): ApiResult<RevokeAttendanceResponse> =
        apiCall(json) { api.revokeAttendance(attendanceId, RevokeAttendanceRequest()) }

    override suspend fun dashboard(): ApiResult<ProfessorDashboardResponse> =
        apiCall(json) { api.dashboard() }

    override suspend fun students(notEnrolledInClassId: String?): ApiResult<List<ProfessorStudent>> =
        apiCall(json) { api.students(notEnrolledInClassId) }.map { it.students }
}
